src/daemon/distribution.c

```diff
@@ -0,0 +1,280 @@
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Svetlana Shmidt <sshmidt at google.com>
+ **/
+
+#include "distribution.h"
+
+#include <errno.h>
+#include <math.h>
+#include <stdio.h>
+#include <string.h>
+
+struct distribution_s {
+  bucket_t *tree;
+  size_t num_buckets;
+  double total_sum;
+};
+
+/**
+ * This code uses an Euler path to avoid gaps in the tree-to-array mapping.
+ * This way the tree contained N buckets contains 2 * N - 1 nodes
+ * Thus, left subtree has 2 * (mid - left + 1) - 1 nodes,
+ * therefore the right subtree starts at node_index + 2 * (mid - left + 1).
+ * For a detailed explanation, see https://docs.google.com/document/d/1ccsg5ffUfqt9-mBDGTymRn8X-9Wk1CuGYeMlRxmxiok/edit?usp=sharing".
+ */
+
+static size_t left_child_index(size_t node_index,
+                               __attribute__((unused)) size_t left,
+                               __attribute__((unused)) size_t right) {
+  return node_index + 1;
+}
+
+static size_t right_child_index(size_t node_index, size_t left, size_t right) {
+  size_t mid = (left + right) / 2;
+  return node_index + 2 * (mid - left + 1);
+}
+
+static size_t tree_size(size_t num_buckets) {
+  return 2 * num_buckets - 1;
+}
+
+static bucket_t merge_buckets(bucket_t left_child, bucket_t right_child) {
+  return (bucket_t) {
+      .bucket_counter = left_child.bucket_counter + right_child.bucket_counter,
+      .maximum = right_child.maximum,
+  };
+}
+
+static void build_tree(distribution_t *d, bucket_t *buckets, size_t node_index, size_t left, size_t right) {
+  if (left > right)
+    return;
+  if (left == right) {
+    d->tree[node_index] = buckets[left];
+    return;
+  }
+  size_t mid = (left + right) / 2;
+  size_t left_child = left_child_index(node_index, left, right);
+  size_t right_child = right_child_index(node_index, left, right);
+  build_tree(d, buckets, left_child, left, mid);
+  build_tree(d, buckets, right_child, mid + 1, right);
+  d->tree[node_index] = merge_buckets(d->tree[left_child], d->tree[right_child]);
+}
+
+static distribution_t* build_distribution_from_bucket_array(size_t num_buckets, bucket_t *bucket_array) {
+  distribution_t *new_distribution = calloc(1, sizeof(*new_distribution));
+  if (num_buckets == 0)
+    return new_distribution;
```

Comment:
I don't understand why you return new_distribution when the num_buckets is equal to zero

Reply:
Maybe it would be better to return NULL in case num_buckets is equal to zero?

Reply:
Thanks for your suggestions. I figured out that this condition is unnecessary because I check the number of buckets in constructors themselves. So I'll just remove it